scipy/optimize/minpack.py

```diff
@@ -145,12 +144,19 @@ def fsolve(func, x0, args=(), fprime=None, full_output=0,
         info['fvec'] = res['fun']
         return x, info, res['status'], res['message']
     else:
+        status = res['status']
+        msg = res['message']
+        if status != 1:
+            if status in [2, 3, 4, 5]:
+                warnings.warn(msg, RuntimeWarning)
+            else:
+                raise TypeError(msg)
```

Comment:
I think this would be clearer with
```
if status == 0:
    raise TypeError(msg)
else:
     warnings.warn(msg, RuntimeWarning)
```

I spent a bit of time looking for `ValueError` that are gone, finally noticing that there were apparently not used at all.